StashRepository: Only consider pull requests in the "OPEN" state

proski opened the pull request:
```
*  StashRepository: Only consider pull requests in the "OPEN" state
   
   Pull requests in other states were wrongly considered buildable.
   
   Add a constructor with trigger and client arguments for unit tests. It
   should be possible to make the code friendlier to unit tests in the
   future.
   
   Add unit tests using Mockito and Hamcrest.
```

proski commented:
Looking through the history, even the oldest revision of this plugin has this bug. On the other hand, bitbucket-pullrequest-builder-plugin only considers pull requests in the "OPEN" state. I believe the bug was introduced when the `shouldBuild` variable was introduced. While open pull requests should be built by default, it's not true for other (e.g declined) pull requests.

I decided to use Mockito because it's a mock framework that doesn't go through backdoors. It encourages writing testable code. Mockito is used in many plugins and in Jenkins itself.

The code is not friendly to unit testing because it uses `new` to create objects without giving an option to supply an existing object. I made a simple change and added a comment to prevent unintended use. It will serve us as a reminder to structure the code better in the future.

The code change reduces indent of a long function, so it may be harder to review. On the positive side, less indented code is easier to understand and to modify.

jakub-bochenski commented:
One thing I don't get is why I can't recall seeing this plugin triggering for PRs that are not open?

> The code change reduces indent of a long function, so it may be harder to review. On the positive side, less indented code is easier to understand and to modify.

Don't worry, if you do "hide whitespace" it's not even visible

proski commented:
> One thing I don't get is why I can't recall seeing this plugin triggering for PRs that are not open?

That's because Stash only returns PRs in the "OPEN" state by default. But I tried adding `state=ALL&` before `start=` in `StashApiClient#pullRequestsPath(int)`, and I saw many PRs in "MERGED" and "DECLINED" states.

The current code works because of a default setting on the the server side. We can use `state=OPEN` in the API call to make it explicit. I just wanted to fix the code that is already there but is not working as intended.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -0,0 +1,101 @@
+    Collection<StashPullRequestResponseValue> buildablePullRequests =
+        stashRepository.getTargetPullRequests();
+    assertThat(buildablePullRequests, hasSize(1));
+    assertThat(buildablePullRequests, hasItems(pullRequest));
```
use `contains()` to do the two assertions at once (it asserts there are no more items)

proski replied:
Done. Thank you!

proski committed:
StashRepository: Change constructor to get needed objects directly

The StashPullRequestsBuilder object is not needed, but the project and
the trigger are used a lot, so pass them to the constructor. Annotate the
constructor arguments with @Nonnull.

No need to defer setting trigger until the init() call, it is not going
to change.

Passing the needed objects directly simplifies mock object injection in
unit tests.

proski commented:
Fixed deprecated code in the unit tests. Changed `StashRepository` constructor to reduce the required setup in unit tests.

proski committed:
StashRepository: Only consider pull requests in the "OPEN" state

Pull requests in other states were wrongly considered buildable.
Fortunately, they are not returned but Stash. Still, the check was in the
code, and it was not working as intended.

Add Mockito based unit tests. Add a constructor that takes the
StashApiClient object, since "new" cannot be mocked.

proski commented:
Rebased. Split the `StashRepository` constructor changes into a separate commit.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -48,16 +50,25 @@
+  }
+
+  // For unit tests only
+  public StashRepository(
```
make it package private?

jakub-bochenski replied:
actually maybe it would be better to just override init method in tests?

proski replied:
I made it package private. I consider annotating it with VisibleForTesting, but that would require a dependency on Google Guava for very little benefit.

I believe overriding `init()` would require making a subclass of StashRepository, which I would rather avoid.

I'm actually going to remove `init()` when adding pipeline support. I think the best testable solution would be to use a class factory, but I'm trying to minimize changes in this PR.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -0,0 +1,92 @@
+
+  @Test
+  public void getTargetPullRequestsSkipsNullStatePullRequests() {
+    StashRepository stashRepository = new StashRepository(project, trigger, stashApiClient);
```
you seem to use the same setup, why not make it an instance field and use a @Before method?

proski replied:
Done. For whatever reason, I started seeing some strange exceptions about missing Jenkins config file, so I added a `JenkinsRule()` call, which resolved them.
